### PR TITLE
[Merged by Bors] - fix(frontends/lean/brackets): parse {a,b,c} as right associative

### PR DESCRIPTION
**Pull request by digama0**
This swaps the order of parsing so that `{a,b,c}` becomes `insert a (insert b (singleton c))` instead of `insert c (insert b (singleton a))`.

### CODE REVIEW

File: src/frontends/lean/brackets.cpp [hunk 1]
```diff
@@ -45,16 +45,24 @@ static expr mk_singleton(parser & p, pos_info const & pos, expr const & e) {
...
         p.next();
         if (p.curr_is_token(get_rcurly_tk())) break;
         expr e2 = p.parse_expr();
-        expr insert = p.save_pos(mk_constant(get_has_insert_insert_name()), ins_pos);
-        r = p.rec_save_pos(mk_app(insert, e2, r), ins_pos);
+        stack.emplace_back(ins_pos, e2);
     }
     p.check_token_next(get_rcurly_tk(), "invalid explicit finite collection, '}' expected");
+    unsigned i = stack.size() - 1;
+    auto [pos, e2] = stack[i];
```
Comment (gebner) [inline src/frontends/lean/brackets.cpp:59]: Lean still uses C++11, so this won't work.

**Comment by urkud** [comment]
Note that with this PR `#check ({0, 1} : set nat)` will output `{1, 0}` because @digama0 changes the parser but not the pretty printer.

File: src/frontends/lean/brackets.cpp (hunk 1 above)
Comment (urkud) [reply src/frontends/lean/brackets.cpp:59]: What's the plan here: rewrite or switch to c++17?

Comment (digama0) [reply src/frontends/lean/brackets.cpp:59]: While I would be fine with switching to C++17, my plan was to rewrite when I get the time to come back to this.

Comment (urkud) [reply src/frontends/lean/brackets.cpp:59]: With `g++ -std=c++17` I get the following error:
```
<...>/lean3/src/frontends/lean/brackets.cpp: In function ‘lean::expr lean::parse_fin_set(lean::parser&, const pos_info&, const lean::expr&)’:
<...>/lean3/src/frontends/lean/brackets.cpp:59:18: error: declaration of ‘auto pos’ shadows a parameter
   59 |     auto [pos, e2] = stack[i];
      |                  ^
make[2]: *** [frontends/lean/CMakeFiles/lean_frontend.dir/build.make:427: frontends/lean/CMakeFiles/lean_frontend.dir/brackets.cpp.o] Error 1
make[1]: *** [CMakeFiles/Makefile2:2226: frontends/lean/CMakeFiles/lean_frontend.dir/all] Error 2
make: *** [Makefile:163: all] Error 2
```

Comment (urkud) [reply src/frontends/lean/brackets.cpp]: Fixed compile. Feel free to revert/rewrite.

**Comment by jcommelin** [comment]
@urkud @digama0 What is the status here? Yury's `has_singleton` PR has been merged. It would be nice to also have this one. And then fix mathlib.

**Comment by gebner** [comment]
The only thing missing here is updating the pretty-printer.

**Comment by urkud** [comment]
Possibly fixed the pretty printer. Compiling & testing.

**Comment by gebner** [comment]
bors merge

**Pull request closed without merging**